File: assets/ops-storyboard.js
/* /ops/ functional-build storyboard. Reads a baked JSON file. Never calls a bus.
 *
 * Entries are date + title + caption + optional illustration. `grain` is
 * "week" or "month". Grouping lives here, so a later bake can summarize to
 * months without a page change.
 */
(function (root, factory) {
  var api = factory();
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  if (root && root.document) api.boot(root);
})(typeof window !== "undefined" ? window : null, function () {
  "use strict";

  var SAME = "/data/ops-storyboard.json";
  var GRAINS = { week: 1, month: 1 };
  var ARTS = { clock: 1, intake: 1, site: 1, voice: 1, honesty: 1, bus: 1, gate: 1, fleet: 1 };
  var DATE_RE = /^\d{4}-\d{2}-\d{2}$/;
  var TS_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;
  var TEXT_RE = /^[A-Za-z0-9][A-Za-z0-9 .,'’+\-—]{0,179}$/;
  var EMAIL_RE = /[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}/i;
  var RETIRED_RE = /foundry|azure/i;
  var BUS_RE = new RegExp(["script" + ".google", "spread" + "sheets", "/" + "macros/", "alpha" + "-db"].join("|"), "i");
  var SECRET_VALUE_RE = new RegExp(
    ["gh" + "p_", "github_" + "pat_", "gho" + "_", "glpat-", "xox[baprs]-", "sk-", "ya29\\.", "AKIA[0-9A-Z]{16}", "-----BEGIN "].join("|"),
    "i"
  );
  var FORBIDDEN = {
    email: 1, "e-mail": 1, mail: 1, token: 1, access_token: 1, refresh_token: 1, id_token: 1,
    secret: 1, password: 1, passwd: 1, credential: 1, credentials: 1, api_key: 1, apikey: 1,
    authorization: 1, auth: 1, transcript: 1, payload: 1, raw: 1, body: 1, cookie: 1,
    cookies: 1, session: 1, private_key: 1, bearer: 1, sha: 1, url: 1
  };
  var CAP = 36;
  var MONTHS = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
  var MONTHS_SHORT = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
  var SCENES = {
    clock: '<circle cx="80" cy="46" r="22"/><path d="M80 46 V30 M80 46 L94 54"/>',
    intake: '<rect x="46" y="28" width="68" height="40" rx="4"/><path d="M46 36 L80 54 L114 36"/>',
    site: '<rect x="38" y="20" width="84" height="52" rx="4"/><path d="M38 34 H122"/><circle cx="46" cy="27" r="2" fill="currentColor" stroke="none"/>',
    voice: '<path d="M58 28 h40 a8 8 0 0 1 8 8 v14 a8 8 0 0 1-8 8 H70 l-12 10 v-10 H58 a8 8 0 0 1-8-8 V36 a8 8 0 0 1 8-8z"/>',
    honesty: '<rect x="50" y="16" width="60" height="58" rx="4"/><path d="M66 46 l8 8 18-20"/>',
    bus: '<circle cx="46" cy="48" r="8"/><circle cx="80" cy="28" r="8"/><circle cx="114" cy="48" r="8"/><path d="M54 44 L74 32 M86 32 L106 44"/>',
    gate: '<path d="M48 62 V28 h64 v34"/><path d="M80 40 v10"/><circle cx="80" cy="36" r="3"/>',
    fleet: '<circle cx="58" cy="40" r="10"/><circle cx="80" cy="52" r="10"/><circle cx="102" cy="40" r="10"/>',
    mark: '<circle cx="80" cy="46" r="6"/>'
  };

  function forbiddenKey(key) {
    if (typeof key !== "string") return true;
    var norm = key.replace(/-/g, "_").toLowerCase();
    return !!FORBIDDEN[norm] || /_token$/.test(norm) || /_secret$/.test(norm);
  }

  function secretish(value) {
    return typeof value === "string" && (EMAIL_RE.test(value) || SECRET_VALUE_RE.test(value) || BUS_RE.test(value));
  }

  function text(value, limit) {
    if (typeof value !== "string") return null;
    var clean = value.replace(/\s+/g, " ").trim();
    if (!clean || clean.length > limit || !TEXT_RE.test(clean) || secretish(clean) || RETIRED_RE.test(clean)) return null;
    return clean;
  }

  function cleanEntry(row) {
    if (!row || typeof row !== "object" || Array.isArray(row)) return null;
    var src = {};
    for (var key in row) {
      if (Object.prototype.hasOwnProperty.call(row, key) && !forbiddenKey(key)) src[key] = row[key];
    }
    if (typeof src.date !== "string" || !DATE_RE.test(src.date)) return null;
    var parts = src.date.split("-");
    var probe = new Date(Date.UTC(+parts[0], +parts[1] - 1, +parts[2]));
    if (probe.getUTCFullYear() !== +parts[0] || probe.getUTCMonth() !== +parts[1] - 1 || probe.getUTCDate() !== +parts[2]) return null;
    var title = text(src.title, 72);
    var caption = text(src.caption, 180);
    if (!title || !caption) return null;
    var out = { date: src.date, title: title, caption: caption };
    if (ARTS[src.illustration]) out.illustration = src.illustration;
    return out;
  }

  function sanitize(raw) {
    if (!raw || typeof raw !== "object" || Array.isArray(raw) || raw.v !== 1) return null;
    if (typeof raw.baked_at !== "string" || !TS_RE.test(raw.baked_at)) return null;
    var grain = GRAINS[raw.grain] ? raw.grain : "week";
    var source = raw.source === "sample" || raw.source === "bake" ? raw.source : "bake";
    var rows = Array.isArray(raw.entries) ? raw.entries : [];
    var entries = [];
    var seen = {};
    for (var i = 0; i < rows.length && entries.length < CAP; i++) {
      var clean = cleanEntry(rows[i]);
      if (!clean) continue;
      var key = clean.date + "\n" + clean.title;
      if (seen[key]) continue;
      seen[key] = 1;
      entries.push(clean);
    }
    entries.sort(function (a, b) { return a.date < b.date ? -1 : a.date > b.date ? 1 : 0; });
    return { v: 1, grain: grain, baked_at: raw.baked_at, source: source, entries: entries };
  }

  function utcDate(iso) {
    var p = iso.split("-");
    return new Date(Date.UTC(+p[0], +p[1] - 1, +p[2]));
  }

  function groupKey(iso, grain) {
    if (grain === "month") return iso.slice(0, 7);
    var d = utcDate(iso);
    var day = d.getUTCDay() || 7;
    var thursday = new Date(d.getTime());
    thursday.setUTCDate(d.getUTCDate() + 4 - day);
    var yearStart = new Date(Date.UTC(thursday.getUTCFullYear(), 0, 1));
    var week = Math.ceil((((thursday - yearStart) / 86400000) + 1) / 7);
    var y = thursday.getUTCFullYear();
    var w = week < 10 ? "0" + week : String(week);
    return y + "-W" + w;
  }

  function groupLabel(iso, grain) {
    var d = utcDate(iso);
    if (grain === "month") return MONTHS[d.getUTCMonth()] + " " + d.getUTCFullYear();
    var day = d.getUTCDay() || 7;
    var monday = new Date(d.getTime());
    monday.setUTCDate(d.getUTCDate() - (day - 1));
    return "Week of " + monday.getUTCDate() + " " + MONTHS_SHORT[monday.getUTCMonth()];
  }

  function groupEntries(entries, grain) {
    var use = grain === "month" ? "month" : "week";
    var groups = [];
    var index = {};
    for (var i = 0; i < entries.length; i++) {
      var row = entries[i];
      var key = groupKey(row.date, use);
      if (!index[key]) {
        index[key] = { key: key, label: groupLabel(row.date, use), entries: [] };
        groups.push(index[key]);
      }
      index[key].entries.push(row);
    }
    return groups;
  }

  function esc(value) {
    return String(value == null ? "" : value).replace(/[&<>"']/g, function (c) {
      return { "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[c];
    });
  }

  function frameDate(iso) {
    var d = utcDate(iso);
    return d.getUTCDate() + " " + MONTHS_SHORT[d.getUTCMonth()];
  }

  function frameSvg(key) {
    var scene = SCENES[key] || SCENES.mark;
    return '<svg viewBox="0 0 160 90" aria-hidden="true" fill="none" stroke="currentColor" stroke-width="1.8" stroke-linecap="round" stroke-linejoin="round">' + scene + "</svg>";
  }

  function paint(snap) {
    if (!snap || !snap.entries || !snap.entries.length) {
      return '<p class="empty">No frames in this storyboard.</p>';
    }
    var grain = snap.grain === "month" ? "month" : "week";
    var groups = groupEntries(snap.entries, grain);
    var html = '<p class="story-meta">' + (grain === "month" ? "Grouped by month" : "Grouped by week") + "</p>";
    for (var g = 0; g < groups.length; g++) {
      var group = groups[g];
      html += '<section class="story-week"><h3>' + esc(group.label) + "</h3>";
      for (var i = 0; i < group.entries.length; i++) {
        var row = group.entries[i];
        html += '<article class="story-frame">' +
          '<div class="frame-art">' + frameSvg(row.illustration) + "</div>" +
          "<div><p class=\"story-kicker\">" + esc(frameDate(row.date)) + "</p>" +
          "<h4>" + esc(row.title) + "</h4>" +
          "<p>" + esc(row.caption) + "</p></div></article>";
      }
      html += "</section>";
    }
    return html;
  }

  function emptyPaint() {
    return '<p class="empty">Storyboard unavailable.</p>';
  }

  return {
    sanitize: sanitize,
    groupEntries: groupEntries,
    groupKey: groupKey,
    paint: paint,
    emptyPaint: emptyPaint,
    esc: esc,
    SAME: SAME,
    boot: boot
  };

  function boot(win) {
    var host = win.document.getElementById("ops-story-mount");
    if (!host) return;

    function show(html) {
      host.innerHTML = html;
    }

    function load() {
      return win.fetch(SAME + "?t=" + Math.floor(Date.now() / 86400000), { cache: "no-store", credentials: "omit" })
        .then(function (res) {
          if (!res.ok) throw new Error("status");
          return res.json();
        })
        .then(function (raw) {
          var snap = sanitize(raw);
          show(snap ? paint(snap) : emptyPaint());
        })
        .catch(function () {
          show(emptyPaint());
        });
    }

    try { load(); } catch (err) { show(emptyPaint()); }
  }
});

File: tools/ops_storyboard.py
#!/usr/bin/env python3
"""Bake the Ops functional-build storyboard. No live bus, no git-log dump.

The page at /ops/ reads data/ops-storyboard.json and draws illustration frames.
Entries are a date, a title, a short caption, and an optional illustration key.
`grain` is "week" now. The same entries group by month when grain is "month",
so a later bake can summarize without a page change.

  python3 tools/ops_storyboard.py --sample --out data/ops-storyboard.json
  python3 tools/ops_storyboard.py --check data/ops-storyboard.json
  python3 tools/ops_storyboard.py --bake --export frames.json --out /tmp/story.json

Weekly cadence starts in .github/workflows/ops-storyboard.yml. That job checks
the file. It does not push to main and it does not call a bus.
"""
from __future__ import annotations

import argparse
import json
import re
import sys
from datetime import date, datetime, timedelta, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_OUT = ROOT / "data" / "ops-storyboard.json"

GRAINS = {"week", "month"}
SOURCES = {"sample", "bake"}
ILLUSTRATIONS = {"clock", "intake", "site", "voice", "honesty", "bus", "gate", "fleet"}
SNAP_KEYS = {"v", "grain", "baked_at", "source", "entries"}
ENTRY_REQUIRED = {"date", "title", "caption"}
ENTRY_KEYS = ENTRY_REQUIRED | {"illustration"}
CAP = 36
TITLE_LIMIT = 72
CAPTION_LIMIT = 180

DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
TS_RE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$")
TEXT_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9 .,'’+\-—]{0,179}$")
EMAIL_RE = re.compile(r"[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}", re.I)
RETIRED_RE = re.compile(r"foundry|azure", re.I)
BUS_RE = re.compile("|".join(("script" + ".google", "spread" + "sheets", "/" + "macros/", "alpha" + "-db")), re.I)
SECRET_VALUE_RE = re.compile(
    "|".join((
        "gh" + "p_",
        "github_" + "pat_",
        "gho" + "_",
        "glpat-",
        "xox[baprs]-",
        "sk-",
        "ya29.",
        "AKIA[0-9A-Z]{16}",
        "-----BEGIN ",
    )),
    re.I,
)
FORBIDDEN_KEYS = {
    "email", "e-mail", "mail", "token", "access_token", "refresh_token", "id_token",
    "secret", "password", "passwd", "credential", "credentials", "api_key", "apikey",
    "authorization", "auth", "transcript", "payload", "raw", "body", "cookie",
    "cookies", "session", "private_key", "bearer", "sha", "url",
}

MONTHS = (
    "January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December",
)
MONTHS_SHORT = (
    "Jan", "Feb", "Mar", "Apr", "May", "Jun",
    "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
)


def _forbidden_key(key) -> bool:
    if not isinstance(key, str):
        return True
    norm = key.strip().lower().replace("-", "_")
    return norm in FORBIDDEN_KEYS or norm.endswith("_token") or norm.endswith("_secret")


def _secretish(value: str) -> bool:
    return bool(EMAIL_RE.search(value) or SECRET_VALUE_RE.search(value) or BUS_RE.search(value))


def _text(value, limit: int) -> str | None:
    if not isinstance(value, str):
        return None
    text = " ".join(value.split())
    if not text or len(text) > limit or not TEXT_RE.match(text):
        return None
    if _secretish(text) or RETIRED_RE.search(text):
        return None
    return text


def _date(value) -> date | None:
    if not isinstance(value, str) or not DATE_RE.match(value):
        return None
    try:
        return date.fromisoformat(value)
    except ValueError:
        return None


def _clean_entry(row) -> dict | None:
    if not isinstance(row, dict):
        return None
    if any(_forbidden_key(k) for k in row):
        row = {k: v for k, v in row.items() if not _forbidden_key(k)}
    day = _date(row.get("date"))
    title = _text(row.get("title"), TITLE_LIMIT)
    caption = _text(row.get("caption"), CAPTION_LIMIT)
    if day is None or not title or not caption:
        return None
    out = {"date": day.isoformat(), "title": title, "caption": caption}
    art = row.get("illustration")
    if isinstance(art, str) and art in ILLUSTRATIONS:
        out["illustration"] = art
    return out


def sanitize(raw, source: str | None = None) -> dict | None:
    """Return an allowlisted storyboard, or None when the envelope is unusable."""
    if not isinstance(raw, dict) or raw.get("v") != 1:
        return None
    if not isinstance(raw.get("baked_at"), str) or not TS_RE.match(raw["baked_at"]):
        return None
    grain = raw.get("grain") if raw.get("grain") in GRAINS else "week"
    src = source if source in SOURCES else raw.get("source")
    if src not in SOURCES:
        src = "bake"
    rows = raw.get("entries") if isinstance(raw.get("entries"), list) else []
    entries = []
    seen = set()
    for row in rows:
        if len(entries) >= CAP:
            break
        clean = _clean_entry(row)
        if not clean:
            continue
        key = (clean["date"], clean["title"])
        if key in seen:
            continue
        seen.add(key)
        entries.append(clean)
    entries.sort(key=lambda row: row["date"])
    return {
        "v": 1,
        "grain": grain,
        "baked_at": raw["baked_at"],
        "source": src,
        "entries": entries,
    }


def problems(snap) -> list[str]:
    if not isinstance(snap, dict):
        return ["storyboard must be an object"]
    if set(snap) != SNAP_KEYS:
        extra = sorted(set(snap) - SNAP_KEYS)
        missing = sorted(SNAP_KEYS - set(snap))
        out = []
        if extra:
            out.append("unknown keys: " + ", ".join(extra))
        if missing:
            out.append("missing keys: " + ", ".join(missing))
        return out
    out = []
    if snap["v"] != 1:
        out.append("v must be 1")
    if snap["grain"] not in GRAINS:
        out.append("grain must be week or month")
    if not isinstance(snap["baked_at"], str) or not TS_RE.match(snap["baked_at"]):
        out.append("baked_at must be UTC like 2026-09-26T06:30:00Z")
    if snap["source"] not in SOURCES:
        out.append("source must be sample or bake")
    entries = snap["entries"]
    if not isinstance(entries, list):
        out.append("entries must be a list")
    elif len(entries) > CAP:
        out.append("entries exceed the cap of %d" % CAP)
    else:
        for i, row in enumerate(entries):
            if not isinstance(row, dict):
                out.append("entries[%d] must be an object" % i)
                continue
            keys = set(row)
            if not ENTRY_REQUIRED <= keys or not keys <= ENTRY_KEYS:
                out.append("entries[%d] keys must stay inside the allowlist" % i)
        if sanitize(snap) != snap:
            out.append("storyboard is not stable under sanitize")
    return out


def monday_of(day: date) -> date:
    return day - timedelta(days=day.weekday())


def group_key(day: date, grain: str) -> str:
    """Stable bucket id. Week is an ISO year-week; month is YYYY-MM."""
    if grain == "month":
        return "%04d-%02d" % (day.year, day.month)
    iso = day.isocalendar()
    return "%04d-W%02d" % (iso.year, iso.week)


def group_label(day: date, grain: str) -> str:
    if grain == "month":
        return "%s %d" % (MONTHS[day.month - 1], day.year)
    start = monday_of(day)
    return "Week of %d %s" % (start.day, MONTHS_SHORT[start.month - 1])


def group_entries(entries, grain: str) -> list[dict]:
    """Oldest first. The page uses the same buckets, so a grain flip needs no rewrite."""
    groups = []
    index = {}
    for row in entries:
        day = date.fromisoformat(row["date"])
        key = group_key(day, grain)
        if key not in index:
            index[key] = {"key": key, "label": group_label(day, grain), "entries": []}
            groups.append(index[key])
        index[key]["entries"].append(row)
    return groups


def _frames() -> list[dict]:
    """Visitor-readable milestones. Not a log, not a commit list."""
    return [
        {
            "date": "2026-08-25",
            "title": "The clock starts",
            "caption": "A working place for the build, ten days before the public site.",
            "illustration": "clock",
        },
        {
            "date": "2026-09-01",
            "title": "A way in",
            "caption": "Messages can arrive and land as something a person can answer.",
            "illustration": "intake",
        },
        {
            "date": "2026-09-04",
            "title": "The site stands up",
            "caption": "The public pages begin. The website is the second half of the story.",
            "illustration": "site",
        },
        {
            "date": "2026-09-06",
            "title": "Someone can be reached",
            "caption": "When the work is blocked, a person can be asked, in one tap.",
            "illustration": "voice",
        },
        {
            "date": "2026-09-18",
            "title": "The page tells the truth",
            "caption": "Words that belonged only in tests come off the live site, and a check keeps them off.",
            "illustration": "honesty",
        },
        {
            "date": "2026-09-26",
            "title": "Ops, in the open",
            "caption": "Specialized agents on a Communication and Control BUS, with a gate between preview and production.",
            "illustration": "bus",
        },
    ]


def sample_story() -> dict:
    snap = sanitize({
        "v": 1,
        "grain": "week",
        "baked_at": "2026-09-26T06:30:00Z",
        "source": "sample",
        "entries": _frames(),
    })
    assert snap is not None
    return snap


def bake(baked_at: str, export, grain: str = "week") -> dict | None:
    raw = export if isinstance(export, dict) else {}
    body = {
        "v": 1,
        "grain": grain if grain in GRAINS else raw.get("grain"),
        "baked_at": baked_at,
        "source": "bake",
        "entries": raw.get("entries") if isinstance(raw.get("entries"), list) else [],
    }
    return sanitize(body, source="bake")


def dumps(snap: dict) -> str:
    return json.dumps(snap, indent=2, sort_keys=True) + "\n"


def validate(path: Path) -> list[str]:
    try:
        snap = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError) as exc:
        return ["not JSON: %s" % exc]
    return problems(snap)


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description="Bake or validate the Ops storyboard")
    ap.add_argument("--out", type=Path, help="Where to write the storyboard")
    ap.add_argument("--sample", action="store_true", help="Write the committed sample")
    ap.add_argument("--check", type=Path, nargs="?", const=DEFAULT_OUT, help="Validate a storyboard file")
    ap.add_argument("--bake", action="store_true", help="Sanitize an export into a storyboard")
    ap.add_argument("--export", type=Path, help="Curated frames JSON, not a log dump")
    ap.add_argument("--grain", choices=sorted(GRAINS), help="week now, month when summarizing")
    ap.add_argument("--baked-at", help="UTC timestamp for tests")
    args = ap.parse_args(argv)
    if args.check:
        bad = validate(args.check)
        for item in bad:
            print(item)
        print("ok" if not bad else "%d problem(s)" % len(bad))
        return 1 if bad else 0
    if not args.out:
        ap.error("--out is required unless --check")
    if args.sample:
        snap = sample_story()
    elif args.bake:
        export = {}
        if args.export:
            try:
                export = json.loads(args.export.read_text(encoding="utf-8"))
            except (OSError, ValueError) as exc:
                print("export unreadable: %s" % exc)
                return 1
            if not isinstance(export, dict):
                print("export must be an object")
                return 1
        stamp = args.baked_at or datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
        grain = args.grain or (export.get("grain") if isinstance(export.get("grain"), str) else "week")
        snap = bake(stamp, export, grain)
        if snap is None:
            print("bake produced nothing renderable")
            return 1
    else:
        ap.error("one of --sample, --bake, or --check is required")
    bad = problems(snap)
    if bad:
        for item in bad:
            print(item)
        return 1
    args.out.parent.mkdir(parents=True, exist_ok=True)
    args.out.write_text(dumps(snap), encoding="utf-8")
    print("wrote %s (%d frames)" % (args.out, len(snap["entries"])))
    return 0


if __name__ == "__main__":
    sys.exit(main())
